# Rede Neural para a Base do Censo Usando Keras
#### Usando a biblioteca do python keras para criar uma rede neural usando a base de dados do censo.

### Pré-processamento da base de dados

In [ ]:
# importando a biblioteca pandas do python
import pandas as pd

# atribuindo os registros da base de dados para o objeto "dataframe"
dataframe = pd.read_csv('census.csv', encoding = 'utf-8', sep = ',')

# separando os atributos previsores e classe 
previsores = dataframe.iloc[:, 0:14].values
classe = dataframe.iloc[:, 14].values

# importando a biblioteca sklearn do python
from sklearn.preprocessing import LabelEncoder
# LabelEncoder é uma função responsável por  normalizar rótulos
# o objetivo é transoformar variáveis categóricas em numéricas

# criando o objeto "labelencoder_previsores" para fazer a transformação dos atributos categóricos em numéricos
labelencoder_previsores = LabelEncoder()

# aplicando a transformação em todos os campos de dados categóricos usando o objeto criado e definido 
# "LabelEncoder_previsores"
previsores[:,1] = labelencoder_previsores.fit_transform(previsores[:,1])
previsores[:,3] = labelencoder_previsores.fit_transform(previsores[:,3])
previsores[:,5] = labelencoder_previsores.fit_transform(previsores[:,5])
previsores[:,6] = labelencoder_previsores.fit_transform(previsores[:,6])
previsores[:,7] = labelencoder_previsores.fit_transform(previsores[:,7])
previsores[:,8] = labelencoder_previsores.fit_transform(previsores[:,8])
previsores[:,9] = labelencoder_previsores.fit_transform(previsores[:,9])
previsores[:,13] = labelencoder_previsores.fit_transform(previsores[:,13])

# criando o objeto 'labelencoder_classe' para fazer a transformação dos atributos categóricos em numéricos
labelencoder_classe = LabelEncoder()

classe = labelencoder_classe.fit_transform(classe)

# importando a biblioteca sklearn do python
from sklearn.preprocessing import StandardScaler

# criando e configurando o objeto "scaler"
scaler = StandardScaler()

# aplicando o escaolonamento nos atributos previsores
previsores = scaler.fit_transform(previsores)

# importando a biblioteca sklearn do python
from sklearn.model_selection import train_test_split
# a função "train_test_split" tem a função de separar modelos de treinamento e modelos de teste em uma base 
#de dados

# separando as bases de dados entre modelos de treinamento e modelos de teste
previsores_treinamento, previsores_teste, classe_treinamento, classe_teste = train_test_split(previsores, 
                                                                                             classe,
                                                                                             test_size = 0.15, 
                                                                                             random_state = 0)

> O pré-processamento acima teve uma boa adapatação com o algoritmo feito. Outros testes foram feitos com pré-processamentos diferentes e os resultados não foram suficientemente bons.

### Usando o Keras para criar uma rede neural

In [2]:
# importando a biblioteca keras do python
import keras

from keras.models import Sequential
# 'Sequential' indica o modelo de rede neural que conecta a camada de neurônios de entrada, ocultas e saída

from keras.layers import Dense
# 'Dense' indica que os neurônios da camada anterior se conectam com os da camada posterior

Using TensorFlow backend.


In [ ]:
# criando o objeto de configuração para a rede neural 'classificador'
classificador = Sequential()

In [4]:
# visualizando a quantidade de dados do dataframe
previsores_treinamento.shape

(27676, 14)

In [ ]:
# adicionando a primeira camada de entrada a rede neural
classificador.add(Dense(units = 8, activation = 'relu', input_dim = 14))

In [ ]:
# adicionando a segunda camada oculta a rede neural
classificador.add(Dense( units = 8, activation = 'relu'))

In [ ]:
# adicionando a quarta camada de saída a rede neural
classificador.add(Dense(units = 1, activation = 'sigmoid'))

In [ ]:
# configurando os parâmetros da rede neural
classificador.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [9]:
# realizando o treinamento com a base de dados
classificador.fit(previsores_treinamento, classe_treinamento, batch_size = 10, nb_epoch = 100)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  


Epoch 1/100
27676/27676 [==============================] - 4s 144us/step - loss: 0.3987 - accuracy: 0.8111
Epoch 2/100
27676/27676 [==============================] - 4s 144us/step - loss: 0.3383 - accuracy: 0.8427
Epoch 3/100
27676/27676 [==============================] - 3s 125us/step - loss: 0.3309 - accuracy: 0.8471
Epoch 4/100
27676/27676 [==============================] - 3s 126us/step - loss: 0.3279 - accuracy: 0.8459
Epoch 5/100
27676/27676 [==============================] - 3s 125us/step - loss: 0.3265 - accuracy: 0.8463
Epoch 6/100
27676/27676 [==============================] - 4s 129us/step - loss: 0.3256 - accuracy: 0.8467
Epoch 7/100
27676/27676 [==============================] - 3s 126us/step - loss: 0.3239 - accuracy: 0.8480
Epoch 8/100
27676/27676 [==============================] - 4s 141us/step - loss: 0.3236 - accuracy: 0.8498
Epoch 9/100
27676/27676 [==============================] - 4s 145us/step - loss: 0.3230 - accuracy: 0.8484
Epoch 10/100
27676/27676 [===========

In [ ]:
# realizando a predição com a base de dados de teste
previsoes = classificador.predict(previsores_teste)

In [11]:
# observando na saída de dados as previsões feitas
print(previsoes)

[[1.4841139e-02]
 [2.1438897e-02]
 [2.4466136e-01]
 ...
 [3.3843826e-04]
 [4.3996010e-02]
 [6.8978620e-01]]


In [12]:
# observando as previsões ideais na saída de dados para a base de teste
print(classe_teste)

[0 0 0 ... 0 0 0]


In [ ]:
# convertendo os registros previstos para uma base binária, com um limiar de 0.5
previsoes = (previsoes > 0.5)

In [14]:
# visualizando os valores previstos pelo algoritmo de forma binarizada
print(previsoes)

[[False]
 [False]
 [False]
 ...
 [False]
 [False]
 [ True]]


### Observando a capacidade de predição do algoritmo

In [ ]:
# importando a biblioteca sklearn do python
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
# visualizado a precisão de acertos do algoritmo
precisao = accuracy_score(classe_teste, previsoes)

In [17]:
# visualizando a precisão de acertos do algoritmo
print(precisao)

0.8452405322415558


O algoritmo obteve **84.52%** de precisão de acertos para a base de dados de acerto.

In [ ]:
# criando a matriz de confusão 
matriz = confusion_matrix(classe_teste, previsoes)

In [19]:
# observando a matriz de confusão
print(matriz)

[[3462  231]
 [ 525  667]]


Observe que houve muitos registros classificado de forma incorreta, mas no geral o algoritmo se saiu bem.

### Visualizando o balanceamento dos dados da classe

In [ ]:
# importando a biblioteca collections do python
import collections

In [21]:
# observando a quantidade de registros para cada classe da base de dados de teste
collections.Counter(classe_teste)

Counter({0: 3693, 1: 1192})

### Alguma dúvida? Entre em contato comigo:

- [Me envie um e-mail](mailto:alysson.barbosa@ee.ufcg.edu.br);